### PR TITLE
Crop oversized gameplay textures instead of downscaling them

### PR DESCRIPTION
**Pull request by frenzibyte**
- Resolves mainly the "incorrect glyph rendering" part of https://github.com/ppy/osu/issues/24973

This is a better approach to handle skins that have certain elements with extra empty space on all sides. Take [this skin](https://osu.ppy.sh/community/forums/topics/1792308?n=1) for example, observe the size of the hitcircle number elements:

CleanShot 2023-09-30 at 01 25 35@2x

The size is already larger than the specified maximum size for hit circles (128x128, which affects the glyph sprites). Cropping the texture from the centre fixes the issue. I've also refactored the max dimensions test scene to output upscaled textures all the way from the `TextureLoaderStore` for the visual test to be accurate.

### CODE REVIEW

File: osu.Game.Tests/Visual/Gameplay/TestScenePlayerMaxDimensions.cs [hunk 1]
```diff
@@ -63,18 +72,66 @@ public UpscaledLegacySkin(IStorageResourceProvider resources)
...
+                }
+
+                private TextureUpload upscale(TextureUpload textureUpload)
+                {
+                    var image = Image.LoadPixelData(textureUpload.Data.ToArray(), textureUpload.Width, textureUpload.Height);
+
+                    // The original texture upload will no longer be returned or used.
+                    textureUpload.Dispose();
+
+                    image.Mutate(i => i.Resize(new Size(textureUpload.Width, textureUpload.Height) * scale_factor));
```
Comment (bdach) [inline osu.Game.Tests/Visual/Gameplay/TestScenePlayerMaxDimensions.cs:127]: This doesn't match the title of the PR. This is not a crop operation, it's a resize still. Which is good, since cropping as I understand it would be _actively_ wrong, but anyhow, what's going on here? Can you explain _precisely_ what's causing the issue this is intending to fix?

Comment (frenzibyte) [reply osu.Game.Tests/Visual/Gameplay/TestScenePlayerMaxDimensions.cs:127]: This is the test scene upscaling legacy skin elements to be tested with the skin's cropping mechanism, it's not the fix of this PR. See the change in `LegacySkinExtensions`.

Also, I stated already what this is supposed to fix in OP.

Comment (bdach) [reply osu.Game.Tests/Visual/Gameplay/TestScenePlayerMaxDimensions.cs:127]: OP says:

> The size is already larger than the specified maximum size for hit circles (128x128, which affects the glyph sprites). Cropping the texture from the centre fixes the issue.

I don't understand how this is supposed to fix what from the screenshot looks like an invalid hit circle number overlap. Can you elaborate on that part?

Also I have a feeling that this is a fix that's only correct to this specific use case and will break many others. I wouldn't be surprised if this is somehow specific to hitcircle numbers.

Comment (frenzibyte) [reply osu.Game.Tests/Visual/Gameplay/TestScenePlayerMaxDimensions.cs:127]: It may seem like it, but the actual reason why it's "broken" is because we're downscaling any glyph in the hit circle's sprite text if it's larger than 128x128 (see [this](https://github.com/ppy/osu/blob/09c9baadec535bed2869d897764d340902868932/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs#L142)). This PR changes the logic to use cropping instead of downscaling.

I can't really think of a case where this is worse than scaling. In fact, I was wondering why I picked scaling in the first place. In osu!(stable), a skin with elements that have extra space on all sides would still appear appropriately in gameplay, as opposed to elements that are upscaled. So cropping should be a better approach to be compatible with stable skins.

Comment (frenzibyte) [reply osu.Game.Tests/Visual/Gameplay/TestScenePlayerMaxDimensions.cs:127]: It appears the change in spacing between hitcircle numbers is occurring due to the following sequence of code snippets:

CleanShot 2023-09-30 at 13 16 53@2x
CleanShot 2023-09-30 at 13 16 30@2x
CleanShot 2023-09-30 at 13 14 32@2x

TL;DR: `LegacySpriteText` uses `texture.ScaleAdjust` to adjust the kerning to be correct. The higher the downscale ratio, the lower the kerning is. Whether that's correct or not, I'm not sure. Enforcing all elements to `ScaleAdjust = 1` show abnormal space between glyphs in gameplay, but I don't want to look at this any further since we haven't been changing `ScaleAdjust` anywhere else besides in `LegacySkin.GetTexture` (and this PR will remove the usage in `WithMaximumSize`), and I don't want to derail this conversation over such behaviour.

**Review by bdach — APPROVED** [review]
- tested skins from issue
- double-checked crop math
- tested a skin with elements smaller than the maximum just to make sure they behave as expected (they do)

seems fine

**Comment by bdach** [comment]
Changes to enforced gameplay bounds based on data from https://github.com/ppy/osu/issues/24973#issuecomment-1747956285 will be applied in a follow-up pull.
